Merge this year's changes for CS 6968 (#24)
pavpanchekha/emberfox: book/forms.md

are used almost everywhere: you fill one out to post on Facebook, to
register to vote, or to search Google.

Rendering widgets
=================

When your browser sends information to a web server, that is usually
information that you've typed into some kind of input area, or a
check-box of some sort that you've checked. So the first step in
communicating with other servers is going to be to draw input areas on
the screen and then allow the user to fill them out.

On the web, there are two kinds of input areas: `<input>` elements,
which are for short, one-line inputs, and `<textarea>` elements, which
are for long, multi-line text. I'll implement `<input>` only, because
because `<textarea>` has a lot of strange properties.[^sig-ws]

> Usually, web browsers communicate with the operating system and ask

Reviewer: Chrome does not do this..

Author: What's the right way to qualify this? "Many browsers..."? Maybe that's not even the right generalization here. Perhaps it should be

> Many applications want their input areas to match input areas for all other applications, so use operating-system libraries to draw an input area directly. That's possible in Tk, but browsers need a lot of control over application styling, so we'll be drawing input areas ourselves.

We can add a footnote saying that Safari and maybe Firefox do still sometimes use OS libraries for input areas, but need to detect when a web style makes that impossible.

Reviewer: New text LGTM

the OS to draw the input areas themselves, because that way the input
areas will match the behavior and appearance of OS input areas. That's
*possible* in Tk,[^1] but in the interests of simplicity we'll be
drawing the input areas ourselves.

[^sig-ws]: Whitespace inside a text area is significant, but text
    still wraps, an unsual combination. Plus, they are pretty similar
    to ordinary `<input>` elements in implementation.
[^1]: In Python, you use the `ttk` library.

`<input>` elements are inline content, like text, laid out in lines.
So to support inputs we'll need a new kind of layout object, which
I'll call `InputLayout`, implemented much like `TextLayout`:

``` {.python .browser}
class InputLayout:
    def __init__(self, node):
        self.node = node
        self.children = []
```

> These `InputLayout` objects need a `layout` method needs to compute

Reviewer: ```suggestion
These `InputLayout` objects need a `layout` method to compute
```

Exercises
=========

> *Check boxes*: Add check boxes. In HTML, check boxes `<input>`

Reviewer: ```suggestion
*Check boxes*: Add checkboxes. In HTML, checkbox `<input>`
```

> elements with the `type` attribute set to `checkbox`. The check box is

Reviewer: ```suggestion
elements with the `type` attribute set to `checkbox`. The checkbox is
```

checked if it has the `checked` attribute set, and unchecked

> otherwise. Submitting check boxes in a form is a little tricky,

Reviewer: ```suggestion
otherwise. Submitting checkboxes in a form is a little tricky,
```

> though. A check box named `foo` only appears in the form encoding if

Reviewer: ```suggestion
though. A checkbox named `foo` only appears in the form encoding if
```
